python/lsst/meas/modelfit/mixture.cc

```diff
@@ -0,0 +1,217 @@
+// -*- lsst-c++ -*-
+/*
+ * LSST Data Management System
+ * Copyright 2008-2013 LSST Corporation.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#include "pybind11/pybind11.h"
+#include "pybind11/stl.h"
+
+#include <sstream>  // Python.h must come before even system headers
+
+#include "numpy/arrayobject.h"
+#include "ndarray/pybind11.h"
+#include "ndarray/eigen.h"
+
+#include "lsst/utils/pybind11.h"
+#include "lsst/meas/modelfit/Mixture.h"
+
+namespace py = pybind11;
+using namespace pybind11::literals;
+
+namespace lsst {
+namespace meas {
+namespace modelfit {
+namespace {
+
+using PyMixtureComponent = py::class_<MixtureComponent>;
+using PyMixtureUpdateRestriction = py::class_<MixtureUpdateRestriction,
+                                              std::shared_ptr<MixtureUpdateRestriction>>;
+using PyMixture = py::class_<Mixture, std::shared_ptr<Mixture>,
+                             afw::table::io::PersistableFacade<Mixture>, afw::table::io::Persistable>;
+
+PyMixtureComponent declareMixtureComponent(py::module & mod) {
+    PyMixtureComponent cls(mod, "MixtureComponent");
+    cls.def("getDimension", &MixtureComponent::getDimension);
+    cls.def_readwrite("weight", &MixtureComponent::weight);
+    cls.def("getMu", &MixtureComponent::getMu);
+    cls.def("setMu", &MixtureComponent::setMu);
+    cls.def("getSigma", &MixtureComponent::getSigma);
+    cls.def("setSigma", &MixtureComponent::setSigma);
+    cls.def(
+        "project",
+        (MixtureComponent (MixtureComponent::*)(int) const)&MixtureComponent::project,
+        "dim"_a
+    );
+    cls.def(
+        "project",
+        (MixtureComponent (MixtureComponent::*)(int, int) const)&MixtureComponent::project,
+        "dim1"_a, "dim2"_a
+    );
+    cls.def(py::init<int>(), "dim"_a);
+    cls.def(py::init<Scalar, Vector const &, Matrix const &>(), "weight"_a, "mu"_a, "sigma"_a);
+    cls.def(
+        "__str__",
+        [](MixtureComponent const & self) {
+            std::ostringstream os;
+            os << self;
+            return os.str();
+        }
+    );
```

Comment:
How about a function `template <class T> toString(T const &) {...}` that uses `<<`? Then it could be freely used for "__str__" and "__repr__". This would fit neatly in `utils/pybind11.h` (though that name needs updating to meet our standards).